From 4c1dc612eeb0b42dfa26559217ccdfb03a4e5ef5 Mon Sep 17 00:00:00 2001
From: Naveen S <quic_naves@quicinc.com>
Date: Fri, 10 May 2024 12:08:37 +0530
Subject: [PATCH] sigma_dut: [WAR] fix interface addr changing for each run
 with MLD

Currently, the main_bssid is used for sigma_dut-ap_0.conf, and an
incremented bssid is used for sigma_dut-ap.conf. However, when starting
hostapd in this sequence, sigma_dut-ap.conf and sigma_dut-ap_0.conf keep
incrementing the main_bssid for each run leading to WFA test failures.

Fix it by storing the permanent MAC address in a file, and then retrieve
and use it for subsequent runs.

Add noscan=1 when ap is dual or number of mld links is more than one
and CoHostedBss is set, to skip scanning so that link 0's address will
be set as mld address.

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Signed-off-by: Naveen S <quic_naves@quicinc.com>
---
 ap.c | 81 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 77 insertions(+), 4 deletions(-)

diff --git a/ap.c b/ap.c
index d2e4c09..5d09a3d 100644
--- a/ap.c
+++ b/ap.c
@@ -112,6 +112,7 @@
 #define IEEE80211_SNAP_LEN_DMG 8
 #endif /* IEEE80211_SNAP_LEN_DMG */
 
+#define SIGMA_DUT_PERM_MAC_ADDR	"/tmp/sigma_perm_mac.txt"
 extern char *sigma_wpas_ctrl;
 extern char *sigma_hapd_ctrl;
 extern char *ap_inet_addr;
@@ -253,6 +254,31 @@ int get_hwaddr(const char *ifname, unsigned char *hwaddr)
 	return 0;
 }
 
+static int set_hw_addr(const char *ifname, unsigned char *addr)
+{
+	struct ifreq ifr;
+	int s;
+
+	s = socket(AF_INET, SOCK_DGRAM, 0);
+	if (s < 0)
+		return -1;
+
+	memset(&ifr, 0, sizeof(ifr));
+
+	strlcpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+
+	ifr.ifr_hwaddr.sa_family = 1;
+	memcpy(ifr.ifr_hwaddr.sa_data, addr, ETH_ALEN);
+
+	if (ioctl(s, SIOCSIFHWADDR, &ifr) < 0) {
+		perror("ioctl");
+		close(s);
+		return -1;
+	}
+	close(s);
+	return 0;
+}
+
 static void ath_ap_set_group_id(struct sigma_dut *dut, const char *ifname,
 				const char *val)
 {
@@ -9435,7 +9461,7 @@ enum sigma_cmd_result cmd_ap_config_commit(struct sigma_dut *dut,
 					   struct sigma_conn *conn,
 					   struct sigma_cmd *cmd)
 {
-	FILE *f;
+	FILE *f, *f1;
 	const char *ifname;
 	char buf[500];
 	char path[100];
@@ -9802,9 +9828,51 @@ write_conf:
 			fclose(f);
 			return -2;
 		}
-	} else if (get_hwaddr(ifname, main_bssid)) {
-		fclose(f);
-		return -2;
+	} else {
+		/* Create a file to store the permanent mac address
+		 * to retain the mac address everytime across the
+		 * app restart/re-run
+		 */
+		if (access(SIGMA_DUT_PERM_MAC_ADDR, F_OK) != 0) {
+			if (get_hwaddr(ifname, main_bssid)) {
+				fclose(f);
+				return ERROR_SEND_STATUS;
+			}
+			f1 = fopen(SIGMA_DUT_PERM_MAC_ADDR, "w");
+			if (f1 == NULL) {
+				sigma_dut_print(dut, DUT_MSG_ERROR,
+						"%s: Failed to open %s",
+						__func__, SIGMA_DUT_PERM_MAC_ADDR);
+				fclose(f);
+				return ERROR_SEND_STATUS;
+			}
+			fprintf(f1, "%02x:%02x:%02x:%02x:%02x:%02x",
+					main_bssid[0], main_bssid[1], main_bssid[2],
+					main_bssid[3], main_bssid[4], main_bssid[5]);
+			fclose(f1);
+		} else {
+			/* Reset hw addr to permanent address */
+			if ((f1 = fopen(SIGMA_DUT_PERM_MAC_ADDR, "r"))) {
+				fscanf(f1, "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+						&main_bssid[0], &main_bssid[1], &main_bssid[2],
+						&main_bssid[3], &main_bssid[4], &main_bssid[5]);
+				fclose(f1);
+				if (set_hw_addr(dut->main_ifname, main_bssid)) {
+					sigma_dut_print(dut, DUT_MSG_ERROR,
+							"%s: Failed to set hw addr %02x:%02x:%02x:%02x:%02x:%02x",
+							__func__,main_bssid[0], main_bssid[1], main_bssid[2],
+							main_bssid[3], main_bssid[4], main_bssid[5]);
+					fclose(f);
+					return ERROR_SEND_STATUS;
+				}
+			} else {
+				sigma_dut_print(dut, DUT_MSG_ERROR,
+						"%s: Failed to open %s in read mode",
+						__func__, SIGMA_DUT_PERM_MAC_ADDR);
+				fclose(f);
+				return ERROR_SEND_STATUS;
+			}
+		}
 	}
 
 	fprintf(f, "bssid=%02x:%02x:%02x:%02x",
@@ -10909,6 +10977,11 @@ skip_vht_parameters_set:
 			fprintf(f, "bssid=%02x:%02x:%02x:%02x:%02x:%02x\n",
 				bssid[0], bssid[1], bssid[2], bssid[3],
 				bssid[4], bssid[5]);
+			/*
+			 * noscan has been set to have MLD address and
+			 * link 0 address same in cohosted BSS case
+			 */
+			fprintf(f, "noscan=1\n");
 
 			if (dut->cohosted[dut->cohost_iface].conf[i].key_mgmt == AP_WPA2_SAE) {
 				fprintf(f, "wpa=2\n");
-- 
2.34.1

